### ISSUE TITLE
Tracking Issue for `Iterator::next_chunk`

### ISSUE THREAD

**Issue by rossmacarthur** [rust-lang/rust#98326]
Feature gate: `#![feature(iter_next_chunk)]`

This is a tracking issue for the `.next_chunk()` method on iterators which allows you to advance the iterator `N` elements and return an array `[T; N]`

### Public API

```rust
use core::array;

trait Iterator {
    type Item;
    
    fn next_chunk<const N: usize>(&mut self,) -> Result<[Self::Item; N], array::IntoIter<Self::Item, N>>
    where
        Self: Sized;
}
```

### Steps / History

- [ ] Implementation: #93700
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Naming: other options include `next_array()` or `next_array_chunk()`.
- Should we also add `next_chunk_back` to `DoubleEndedIterator`?

**Comment by ChayimFriedman2** [comment]
Should this method support `N=0`?  [`itertools` does not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e63d3e044909f8a0d73b7da821f2a772), and it sounds plausible; however, if we want to do the same, we will have the same concern as `<[T]>::array_chunks()` that cannot be stabilized until `generic_const_exprs` or some subset will be stabilized.

**Comment by the8472** [comment]
I'm wondering about the intended uses of this API.

If it is for performance (autovectorization) then it seems fairly brittle. I just tried doing this benchmark:

```rust
#[bench]
fn bench_next_chunk(b: &mut Bencher) {
    let v = vec![13u8; 2048];

    b.iter(|| {
        const CHUNK: usize = 8;

        let mut sum = [0u32; CHUNK];
        let mut iter = black_box(v.clone()).into_iter();

        while let Ok(chunk) = iter.next_chunk::<CHUNK>() {
            for i in 0..CHUNK {
                sum[i] += chunk[i] as u32;
            }
        }

        sum
    })
}
```

and it took `17,272 ns/iter` when compiled with `target-cpu=znver2`. With a handrolled implementation I got `211 ns/iter`. I  encountered some 10x slowdowns in a few other combinations. Sometimes `codegen-units=1` produced even worse assembly because vectorizer did silly things such as loading one byte at a time into a simd register with `vpinsrb`.

And that's with direct, linear memory access on a `Vec::IntoIter` which should be relatively straight-forward to turn into masked simd loads. Many iterators yield references. A `[&T; N]` would need a gather-load which is expensive and only available on newer CPUs, this is unlike `slice::array_chunks` which yields `&[T; N]`. We could try specializing `slice.iter().next_chunk()` but then you might as well use `array_chunks`. And of course it'll get more complex with additional adapters.

Another limitation is that it can't be used in a for-in loop or chained further with other iterator adapters, an issue #92393 didn't have.

**Comment by Lokathor** [comment]
This is probably an example of a case where you should first use `copied()`, and then chunk it, and then you turn that array into a simd type for the work, before possibly turning it back at the end of your loop.

And the change into and out of simd form will have a cost, so you'll want to ensure you're doing enough work inside each loop pass or it might exceed the cost of the scalar only code.

**Comment by the8472** [comment]
yeah that was just an example to illustrate that the default implementation is suboptimal and we'll need optimized implementations on most linear-memory sources and length-preserving adapters.

**Comment by Lokathor** [comment]
Mm, yeah.

Myself I would imagine using this most on iterators that *aren't* linear memory sources, and "the point" would be to have a clean and unified way to turn some crazy data source into simple SIMD chunks.

**Comment by rossmacarthur** [comment]
Might be good to align the naming of this function with `Iterator::next_chunk`, i.e. either 

- `Iterator::chunks`
- `Iterator::next_chunk`

Or

- `Iterator::array_chunks`
- `Iterator::next_array_chunk`

**Comment by mark-i-m** [comment]
So I'm finding myself caring less and less about the actual naming. I've wanted this feature at least 3 times in the last year and had to write my own version to get around. Can we just pick something and get the feature stabilized?

**Comment by ZhennanWu** [comment]
I believe the current implementation also suffers from this compiler optimization issue: https://github.com/rust-lang/rust/issues/110734

Godbolt: https://godbolt.org/z/Te5Wo9eh7

**Comment by the8472** [comment]
I'm not sure what that issue has to do with `next_chunk` specifically. It seems to be mostly about the codegen for `it.collect::<Vec<_>>()`, which I think is easily disturbed by _anything_ mutating the iterator.

**Comment by pitaj** [comment]
I'd like to move this more toward specialization. Here are my answers to the Unresolved Questions:

> Naming: other options include `next_array()` or `next_array_chunk()`.

`next_chunk` seems like the best option to me.

> Should we also add `next_chunk_back` to `DoubleEndedIterator`?

Probably, but that should be under a separate feature and not block this.

> How should we handle N = 0?

It should be allowed and return a zero-length array without advancing the iterator. No good reason to forbid it.

**Comment by the8472** [comment]
> It should be allowed and return a zero-length array without advancing the iterator. No good reason to forbid it.

For `next_chunk` it might not be a problem but for the chunks iterators on slices and for iterator.array_chunks() it is. So we might as well do something about it here

**Comment by pitaj** [comment]
I disagree. For those, it's necessary by definition. There's no reason to apply the same logic here besides, in my opinion, a misguided desire for consistency. The behavior of `next_chunk` is totally reasonable for array length 0, unlike the others you mentioned.

"Doing something about it" will pretty much require const expressions just like all the other ones, needlessly delaying the stabilization.

**Comment by the8472** [comment]
Well, it would prevent implementations of `next_chunk` using those things internally

**Comment by pitaj** [comment]
I can see that maybe being the case down the line unless we add something like `const if` or `const match`. But currently they just panic when  N=0 so one can just put a simple

```
if N == 0 {
    return Ok([]);
}
```

at the top which will easily get optimized out.

**Comment by frewsxcv** [comment]
@rust-lang/libs Is there anything blocking a stabilization pull request? I'm happy to open that

**Comment by Andlon** [comment]
I think for the name `next_chunk` it's slightly unfortunate that a "chunk" in this context is an array and not a slice, like `slice.chunks()`. This seems inconsistent and a clear potential source of confusion. `next_array_chunk` or similar removes any such ambiguities, and should be preferred, in my opinion.

If things are otherwise blocked on the handling of `N == 0`, would it be possible to delay this decision by panicking (at compile time?) for now, and then possibly allow it in the future? My understanding is that the case `N == 0` is an edge case that would only come up in some unusual situations involving generics, and it seems unfortunate for such an edge case to delay stabilization for the common use cases of this method.

**Comment by the8472** [comment]
> I think for the name next_chunk it's slightly unfortunate that a "chunk" in this context is an array and not a slice, like slice.chunks(). This seems inconsistent and a clear potential source of confusion. next_array_chunk or similar removes any such ambiguities, and should be preferred, in my opinion.

Whether a chunk is a slice or an array already is not consistent. E.g. slices already have `as_chunks` that returns a reference to arrays. I don't see much of a chance for confusion because `-> [Self::Item]` signatures are not possible and `-> &[Self::Item]` signatures would be neither practical nor useful to implement.

> If things are otherwise blocked on the handling of N == 0, would it be possible to delay this decision by panicking (at compile time?) for now, and then possibly allow it in the future?

That would block the option of adding a `where  N > 0` bound once that's possible

**Comment by Andlon** [comment]
> Whether a chunk is a slice or an array already is not consistent. E.g. slices already have `as_chunks` that returns a reference to arrays. I don't see much of a chance for confusion because `-> [Self::Item]` signatures are not possible and `-> &[Self::Item]` signatures would be neither practical nor useful to implement.

Fair point, though I don't think any of these methods have been stabilized, unless I'm mistaken. `as_chunks` would probably also be better as `as_array_chunks`, for the same reason.

Actually, upon reviewing this post I see that `last_chunk` was recently made available in `1.77`. Unfortunate. Oh well. In that case, you are correct, and that it's already unfortunately inconsistent.

In any case, the confusion I point to has to do with readability. By using clear and distinct naming, you do not have to stop and think through a logical argument like the one you suggest when scanning over the code.

I just don't see any downside in using a few more characters to avoid the ambiguity. But I also don't think we need to bikeshed this further.

> That would block the option of adding a `where N > 0` bound once that's possible

Thanks for pointing this out, that's what I was missing.

**Comment by betelgeuse** [comment]
Before making this stable it would be useful to improve the documentation so that it provides an example with a loop. Now the examples have fixed amount of calls to next_chunk. I would think the more common use case is when not operating on a known input.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.next_chunk

**Comment by jmevel** [comment]
Hi guys,

Just wondering if any of you also experienced some stack overflow errors using this feature?
I'm trying to use it (inside a Tokio task) but whenever I try to get a chunk of `475` or more element I get a stack overflow error. This is fine up to 474 elements.

I'm iterating over huge CSV files using the `csv` crate. I wanted to simplify a bit my code which was calling `next()` on my iterator and executing some code every 10_000 elements but I think I'll have to stick to my existing working code for now.

Just wanted to know if anyone else also stumbled on this issue?

Thanks

**Comment by SuperSamus** [comment]
The generated assembly is pretty bad.
[Godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:37,positionColumn:1,positionLineNumber:37,selectionStartColumn:1,selectionStartLineNumber:37,startColumn:1,startLineNumber:37),source:'%23!!%5Bfeature(iter_next_chunk)%5D%0A%0A%23%5Bunsafe(no_mangle)%5D%0Apub+fn+simd_sum_slow(arr:+%26%5Bu32%5D)+-%3E+u32%0A%7B%0A++++const+STEP_SIZE:+usize+%3D+16%3B%0A%0A++++let+mut+result+%3D+%5B0%3B+STEP_SIZE%5D%3B%0A%0A++++let+mut+iter+%3D+arr.iter()%3B%0A%0A++++while+let+Ok(c)+%3D+iter.next_chunk::%3CSTEP_SIZE%3E()+%7B%0A++++++++for+(%26n,+r)+in+c.iter().zip(result.iter_mut())+%7B%0A++++++++++++*r+%2B%3D+n%3B%0A++++++++%7D%0A++++%7D%0A%0A++++result.iter().sum()%0A%7D%0A%0A%23%5Bunsafe(no_mangle)%5D%0Apub+fn+simd_sum_fast(arr:+%26%5Bu32%5D)+-%3E+u32%0A%7B%0A++++const+STEP_SIZE:+usize+%3D+16%3B%0A%0A++++let+mut+result+%3D+%5B0%3B+STEP_SIZE%5D%3B%0A%0A++++for+c+in+0..(arr.len()+/+STEP_SIZE)+%7B%0A++++++++for+i+in+0..STEP_SIZE+%7B%0A++++++++++++//+%60.get_unchecked()+needed+because+of+https://github.com/rust-lang/rust/issues/134636%0A++++++++++++result%5Bi%5D+%2B%3D+unsafe+%7B*arr.get_unchecked(c*STEP_SIZE+%2B+i)%7D%3B%0A++++++++%7D%0A++++%7D%0A%0A++++result.iter().sum()%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3+-C+target-cpu%3Dx86-64-v3',overrides:!((name:edition,value:'2024')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+nightly+(Editor+%231)',t:'0'),(h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:nightly,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'',overrides:!(),runtimeTools:!(),source:1,stdinPanelShown:'1',wrap:'1'),l:'5',n:'0',o:'Executor+rustc+nightly+(Rust,+Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Benchmarking on my PC, the one that doesn't use `next_chunk` is almost 3 times as fast.

EDIT: Still an issue on 20/08/2026.

**Comment by simonsso** [comment]
The use of Result return type seems a bit out of place since the Err is not a regular `Error` as one could expect. I think it would make more sense to return an enum with the states `Chunk([u8;32])` and `Remainder(iter)` instead

**Comment by max-heller** [comment]
> The use of Result return type seems a bit out of place since the Err is not a regular `Error` as one could expect. I think it would make more sense to return an enum with the states `Chunk([u8;32])` and `Remainder(iter)` instead

The `Result`-based signature feels similar to [`slice::binary_search() -> Result<usize, usize>`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) and `slice::contains() -> bool`

**Comment by asder8215** [comment]
Are we able to add `DoubleEndedIterator::next_chunk_back` underneath this tracking issue? The person who made the [ACP](https://github.com/rust-lang/libs-team/issues/734) for this method and the [original PR](https://github.com/rust-lang/rust/pull/151668) had it marked under `#![feature(iter_next_chunk)]` feature gated. I picked up the PR [here](https://github.com/rust-lang/rust/pull/156737) to get it over the finish line due to inactivity.

**Comment by aapoalas** [comment]
Added `.next_chunk_back()` to the tracking issue.

**Comment by asder8215** [comment]
@aapoalas to note, `.next_chunk_back()` should be underneath the `DoubleEndedIterator` trait not `Iterator` trait.

**Comment by giang299** [comment]
> * Naming: other options include `next_array()` or `next_array_chunk()`.

IT MUST BE next_chunk()
next_chunk is amz naming. we already have a Vec::chunks(),..etc before, it's universal.
